pkg/deploy/tls.go

```diff
@@ -0,0 +1,448 @@
+		if role == nil {
+			return reconcile.Result{RequeueAfter: time.Second}, err
+		}
+
+		roleBiding, err := SyncRoleBindingToCluster(checluster, CheTLSJobRoleBindingName, CheTLSJobServiceAccountName, CheTLSJobRoleName, "Role", clusterAPI)
+		if roleBiding == nil {
+			return reconcile.Result{RequeueAfter: time.Second}, err
+		}
+
+		cheTLSSecretsCreationJobImage := DefaultCheTLSSecretsCreationJobImage()
+		jobEnvVars := map[string]string{
+			"DOMAIN":                         checluster.Spec.K8s.IngressDomain,
+			"CHE_NAMESPACE":                  checluster.Namespace,
+			"CHE_SERVER_TLS_SECRET_NAME":     cheTLSSecretName,
+			"CHE_CA_CERTIFICATE_SECRET_NAME": CheTLSSelfSignedCertificateSecretName,
+		}
+		job, err := SyncJobToCluster(checluster, CheTLSJobName, CheTLSJobComponentName, cheTLSSecretsCreationJobImage, CheTLSJobServiceAccountName, jobEnvVars, clusterAPI)
+		if err != nil {
+			logrus.Error(err)
+			return reconcile.Result{RequeueAfter: time.Second}, err
+		}
+		if job == nil || job.Status.Succeeded == 0 {
+			logrus.Infof("Waiting on job '%s' to be finished", CheTLSJobName)
+			return reconcile.Result{RequeueAfter: time.Second}, err
+		}
+	}
+
+	// cleanup job
+	job := &batchv1.Job{}
+	err = clusterAPI.Client.Get(context.TODO(), types.NamespacedName{Name: CheTLSJobName, Namespace: checluster.Namespace}, job)
+	if err != nil && !errors.IsNotFound(err) {
+		// Failed to get the job
+		return reconcile.Result{RequeueAfter: time.Second}, err
+	}
+	if err == nil {
+		// The job object is present
+		if job.Status.Succeeded > 0 {
+			logrus.Infof("Import public part of Eclipse Che self-signed CA certificvate from \"%s\" secret into your browser.", CheTLSSelfSignedCertificateSecretName)
+			deleteJob(job, checluster, clusterAPI)
+		} else if job.Status.Failed > 0 {
+			// The job failed, but the certificate is present, shouldn't happen
+			deleteJob(job, checluster, clusterAPI)
+			return reconcile.Result{}, nil
+		}
+		// Job hasn't reported finished status yet, wait more
+		return reconcile.Result{RequeueAfter: time.Second}, nil
+	}
+
+	// Che TLS certificate exists, check for required data fields
+	if !isCheTLSSecretValid(cheTLSSecret) {
+		// The secret is invalid because required field(s) missing.
+		logrus.Infof("Che TLS secret \"%s\" is invalid. Recreating...", cheTLSSecretName)
+		// Delete old invalid secret
+		if err = clusterAPI.Client.Delete(context.TODO(), cheTLSSecret); err != nil {
+			logrus.Errorf("Error deleting Che TLS secret \"%s\": %v", cheTLSSecretName, err)
+			return reconcile.Result{RequeueAfter: time.Second}, err
+		}
+		// Recreate the secret
+		return reconcile.Result{RequeueAfter: time.Second}, err
+	}
+
+	// Check owner reference
+	if cheTLSSecret.ObjectMeta.OwnerReferences == nil {
+		// Set owner Che cluster as Che TLS secret owner
+		if err := controllerutil.SetControllerReference(checluster, cheTLSSecret, clusterAPI.Scheme); err != nil {
```

Comment:
Is it better to move this code to `getSpecSecret`
We set controllerRefence for all components at getSpec func

Reply:
The same as above